### ISSUE TITLE
AdaptiveAvgPool3d produces incorrect results/NAN/INF  under torch AMP or when inputs are HALF

### ISSUE THREAD

**Issue by zmy1116** [pytorch/pytorch#52719]
## 🐛 Bug

This is an extremely weird issue. I'm trying to do training on video dataset using torch amp. However, I notice that frequently the model produces NAN loss. After some investigation, I pin the issue to `AdaptiveAvgPool3d `: 

- It looks like for the same input, `AdaptiveAvgPool3d` produces incorrect result if the inputs is casted to half 
- For some specific inputs, `AdaptiveAvgPool3d` produces INF/NAN under torch.amp

 
## To Reproduce

To reproduce inconsistency when inputs in HALF
```
import torch
m = torch.nn.AdaptiveAvgPool3d((1,1,1))

inputs = torch.rand([8,54,16,56,56])
inputs = inputs.cuda()
inputs_2 = inputs.half()

out = m(inputs)
out2 = m(inputs_2)

print('Discepancies', torch.sum(torch.abs(out2- out)) )
```
You should see a large discrepancy, what i see is:
```
Discepancies tensor(198.3446, device='cuda:0')
```

To creating INF output when using `torch.amp`

Download the following `example_data.pt` 
https://drive.google.com/file/d/1XYIZsDzSkRnpLM-FSUyBRAQwUTx2MDP4/view?usp=sharing

```
import torch
data = torch.load('example_data.pt')
m = torch.nn.AdaptiveAvgPool3d((1,1,1))
with torch.cuda.amp.autocast():
    out = m(data)
print(torch.max(out))
```

You should see 
```
tensor(inf, device='cuda:0', dtype=torch.float16, grad_fn=<MaxBackward1>)
```

## Expected behavior
- consistent and correct result for `AdaptiveAvgPool3d` under `torch.amp` or when inputs in fp16 vs fp32
 

## Environment
```
PyTorch version: 1.7.1
Is debug build: False
CUDA used to build PyTorch: 11.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 16.04.7 LTS (x86_64)
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609
Clang version: Could not collect
CMake version: version 3.19.4

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.0.130
GPU models and configuration: GPU 0: Tesla V100-SXM2-16GB
Nvidia driver version: 450.80.02
cuDNN version: Probably one of the following:
/usr/local/cuda-10.1/targets/x86_64-linux/lib/libcudnn.so.7.6.5
/usr/local/cuda-10.2/targets/x86_64-linux/lib/libcudnn.so.7.6.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.0.5
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.0.5
/usr/local/cuda-11.1/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.7.1
[pip3] torchaudio==0.7.0a0+a853dff
[pip3] torchvision==0.8.2
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               11.0.221             h6bb024c_0
[conda] magma-cuda110             2.5.2                         1    pytorch
[conda] mkl                       2020.2                      256
[conda] mkl-include               2020.2                      256
[conda] mkl-service               2.3.0            py37he8ac12f_0
[conda] mkl_fft                   1.2.0            py37h23d657b_0
[conda] mkl_random                1.1.1            py37h0573a6f_0
[conda] numpy                     1.19.2           py37h54aff64_0
[conda] numpy-base                1.19.2           py37hfa32c7d_0
[conda] pytorch                   1.7.1           py3.7_cuda11.0.221_cudnn8.0.5_0    pytorch
[conda] torchaudio                0.7.2                      py37    pytorch
[conda] torchvision               0.8.2                py37_cu110    pytorch

```
## Additional context

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @ngimel @heitorschueroff @albanD @mruberry

**Comment by ngimel** [comment]
High priority for silent wrong results. 
Unlike 2d adaptive pooling, 3d adaptive pooling does not have a shortcut to compute mean when output size is (1,1,1), and instead goes to a specialized kernel that accumulates in half instead of fp32.

**Comment by ngimel** [comment]
@zmy1116 as a workaround, instead of AdaptiveAvgPool3d you can use `x.mean(dim={2,3,4})`

**Comment by zmy1116** [comment]
> @zmy1116 as a workaround, instead of AdaptiveAvgPool3d you can use `x.mean(dim={2,3,4})`

Thanks @ngimel , indeed it works
